ic-factory/src/state.rs

```diff
@@ -1,139 +1,455 @@
-            icp_to,
-            controller,
-        }
-    }
+#[derive(Debug, CandidType, Deserialize)]
+pub struct CanisterModule {
+    /// The canister wasm.
+    wasm: Vec<u8>,
+    /// Canister wasm hash.
+    hash: CanisterHash,
+    /// Canister state version.
+    version: u32,
+    /// Candid-serialized definition of the canister state type.
+    state_header: CandidHeader,
 }
 
-impl Default for FactoryConfiguration {
-    fn default() -> Self {
-        Self {
-            ledger_principal: Principal::anonymous(),
-            icp_fee: DEFAULT_ICP_FEE,
-            icp_to: Principal::anonymous(),
-            controller: Principal::anonymous(),
-        }
+impl CanisterModule {
+    pub fn hash(&self) -> &Vec<u8> {
+        &self.hash
     }
-}
 
-#[derive(CandidType, Deserialize, IcStorage)]
-pub struct FactoryState {
-    pub configuration: FactoryConfiguration,
-    pub factory: Factory,
+    pub fn version(&self) -> u32 {
+        self.version
+    }
 }
 
 impl Versioned for FactoryState {
-    type Previous = ();
+    type Previous = FactoryStateV1;
 
-    fn upgrade((): ()) -> Self {
-        Self::default()
-    }
-}
+    fn upgrade(prev: Self::Previous) -> Self {
+        let FactoryStateV1 {
+            configuration,
+            factory,
+        } = prev;
+        let Factory {
+            canisters,
+            checksum,
+        } = factory;
+
+        let hash = checksum.hash;
 
-impl Default for FactoryState {
-    fn default() -> Self {
         Self {
-            configuration: FactoryConfiguration::default(),
-            factory: Factory::new(&[]),
+            configuration,
+
+            // After the upgrade the canister wasm module would have to be uploaded again to
+            // provide the state header.
+            upgrading_module: None,
+
+            // We assume for now that the canisters were not modified by external controllers, as
+            // we didn't keep track of each canister has before.
+            canisters: canisters
+                .into_iter()
+                .map(|(principal, _)| (principal, hash.clone()))
+                .collect(),
+
+            update_lock: UpdateLock::default(),
         }
     }
 }
 
-/// This trait must be implemented by a factory state to make using of `init_factory_api` macro
-/// possible.
 impl FactoryState {
-    pub fn ledger_principal(&self) -> Principal {
-        self.configuration.ledger_principal
+    pub fn new(configuration: FactoryConfiguration) -> Self {
+        Self {
+            configuration,
+            ..Default::default()
+        }
+    }
+
+    /// Checks if the request caller is the factory conctroller (owner).
+    ///
+    /// # Errors
+    ///
+    /// Returns `FactoryError::AccessDenied` if the caller is not the factory controller.
+    pub fn authorize_owner(&mut self) -> Result<Authorized<Owner>, FactoryError> {
+        let caller = ic_canister::ic_kit::ic::caller();
+        self.authorize_owner_int(caller)
```

Comment:
I can't see a reason for `authorize_owner_int`, it might as well all be in this function